webrender/src/mask_cache.rs

```diff
@@ -8,158 +8,273 @@ use prim_store::{CLIP_DATA_GPU_SIZE, MASK_DATA_GPU_SIZE};
 use renderer::VertexDataStore;
 use util::{MatrixHelpers, TransformedRect};
 use webrender_traits::{AuxiliaryLists, BorderRadius, ClipRegion, ComplexClipRegion, ImageMask};
-use webrender_traits::{DeviceIntRect, DeviceIntSize, LayerRect, LayerToWorldTransform};
+use webrender_traits::{DeviceIntRect, LayerToWorldTransform};
+use webrender_traits::{LayerRect, LayerPoint, LayerSize};
+
+const MAX_CLIP: f32 = 1000000.0;
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ClipMode {
+    Clip,           // Pixels inside the region are visible.
+    ClipOut,        // Pixels outside the region are visible.
+}
 
 #[derive(Clone, Debug)]
 pub enum ClipSource {
-    NoClip,
-    Complex(LayerRect, f32),
-    Region(ClipRegion),
+    Complex(LayerRect, f32, ClipMode),
+    // The bool here specifies whether to consider the rect
+    // from the clip region as part of the mask. This is true
+    // for clip/scroll nodes, but false for primitives, where
+    // the clip rect is handled in local space.
+    Region(ClipRegion, bool),
 }
 
 impl ClipSource {
     pub fn image_mask(&self) -> Option<ImageMask> {
         match self {
-            &ClipSource::NoClip => None,
             &ClipSource::Complex(..) => None,
-            &ClipSource::Region(ref region) => region.image_mask,
+            &ClipSource::Region(ref region, _) => region.image_mask,
         }
     }
 }
 
 #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
 pub struct ClipAddressRange {
     pub start: GpuStoreAddress,
-    item_count: u32,
+    item_count: usize,
+}
+
+/// Represents a local rect and a device space
+/// bounding rect that can be updated when the
+/// transform changes.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Geometry {
+    pub local_rect: LayerRect,
+    pub bounding_rect: DeviceIntRect,
+}
+
+impl Geometry {
+    fn new(local_rect: LayerRect) -> Geometry {
+        Geometry {
+            local_rect: local_rect,
+            bounding_rect: DeviceIntRect::zero(),
+        }
+    }
+
+    fn update(&mut self,
+              transform: &LayerToWorldTransform,
+              device_pixel_ratio: f32) {
+        let transformed = TransformedRect::new(&self.local_rect,
+                                               &transform,
+                                               device_pixel_ratio);
+        self.bounding_rect = transformed.bounding_rect;
+    }
+}
+
+/// Depending on the complexity of the clip, we may either
+/// know the outer and/or inner rect, or neither or these.
+/// In the case of a clip-out, we currently set the mask
+/// bounds to be unknown. This is conservative, but ensures
+/// correctness. In the future we can make this a lot
+/// more clever with some proper region handling.
+#[derive(Clone, Debug, PartialEq)]
+pub enum MaskBounds {
+    /// We haven't calculated the bounds yet.
+    Unknown,
```

Comment:
I think this should be idiomatically used as `Option<MaskBounds>` (instead of `MaskBounds::Unknown`)